From a4f7bac826435d8ee22970e8561e9a1931cc4784 Mon Sep 17 00:00:00 2001
From: Daniel Girardeau-Montaut <daniel.girardeau@gmail.com>
Date: Sun, 9 Jul 2017 10:58:44 +0200
Subject: [PATCH] CMake script syntax fix ('elseif' instead of 'else')

---
 CMakeLists.txt | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6ef646c1..25424b33 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -67,13 +67,13 @@ if( WIN32 )
     if( ${OPTION_BUILD_CCVIEWER} )
         list( APPEND INSTALL_DESTINATIONS ${CCVIEWER_DEST_FOLDER} )
     endif()
-else( APPLE )
+elseif( APPLE )
 	set( INSTALL_DESTINATIONS "${CLOUDCOMPARE_MAC_FRAMEWORK_DIR}" )
     
     if( ${OPTION_BUILD_CCVIEWER} )
         list( APPEND INSTALL_DESTINATIONS "${CCVIEWER_MAC_FRAMEWORK_DIR}" )
     endif()
-else( UNIX )
+elseif( UNIX )
   set( INSTALL_DESTINATIONS ${CMAKE_INSTALL_PREFIX}/bin )
 endif()
 
